### DISCUSSION TITLE
Filtering cascading table widgets

### DISCUSSION THREAD

**Discussion by Bitello** [Budibase/budibase#9033] [Help] [answered]
Hello guys!

I've exhausted my attempts to figure out how to solve a functional requirement on my own.

This is the requirement:
- Table A has a 1:n relationship with Table B which has a 1:n relationship with Table C.

- On one page, I need to display the 3 tables (using the table widget).

- When loading the page, all table widgets will be loaded with all data from their respective tables.

- If I select a row or more from Table A, Table B must be filtered by the corresponding foreign keys, and the same must happen with Table C, which will be filtered by the foreign keys with Table B.

- If I select one or more rows from Table B, Table C should be filtered by the corresponding foreign keys.

My problem:
Hello guys!

I've exhausted my attempts to figure out how to solve a functional requirement on my own.

This is the requirement:
- Table A has a 1:n relationship with Table B which has a 1:n relationship with Table C.

- On one page, I need to display the 3 tables (using the table widget).

- When loading the page, all table widgets will be loaded with all data from their respective tables.

- If I select a row or more from Table A, Table B must be filtered by the corresponding foreign keys, and the same must happen with Table C, which will be filtered by the foreign keys with Table B.

- If I select one or more rows from Table B, Table C should be filtered by the corresponding foreign keys.

My problem:
- I don't know how, when selecting a row from table A, update subsequent tables.

Could someone guide me?

Grateful

**Comment by andz-bb** [comment, marked as answer]
hey @Bitello 

I built an example of this for you:
![table filtering](https://user-images.githubusercontent.com/110921612/207862662-69a5bbb2-4a41-4b90-8e52-ec7cbe98bcd6.gif)


In order to enable this sort of filtering with Budibase DB relationships, you need to create an extra column on each child table that holds the _id of the related parent row. This is done by adding using a static formula that returns the _id from the relationship column like so:
<img width="417" alt="static formula with relationship id" src="https://user-images.githubusercontent.com/110921612/207863452-2bdc32cf-664b-495c-b1ee-077e88fcf188.png">

In this example my table A is Employees, my table B is Jobs and my table C is Subtasks. In the above screenshot this is showing a static formula column I set up in my Jobs table which returns the Employee _id for each job. I can then filter my columns using this value.

I'm using standard table blocks and then adding an action on row click to set the selected row _id to the state.

<img width="821" alt="set row id on state" src="https://user-images.githubusercontent.com/110921612/207864244-23021667-a9b7-470f-a0b4-86b832c8dba0.png">

I then use that value from state to filter the data on the table below.

<img width="1093" alt="filter table based on state id" src="https://user-images.githubusercontent.com/110921612/207864541-38e76337-d30b-40ba-a921-7f37ba4d8449.png">

[three tables-export-1671109053798.tar.gz](https://github.com/Budibase/budibase/files/10237137/three.tables-export-1671109053798.tar.gz)



I might have missed something here in my explanation so if there's anything that's unclear let me know!  I've also attached the app export for you to try out and have a closer look. I hope this is helpful for you! 😊

**Comment by Bitello** [comment]
Hey, @andz-bb 

Thanks for the solution, it was exactly what I was looking for.

But I still have a problem...

Loading your solution file, it works perfectly.
When I tried to implement it in my application, I couldn't use the "Update State" action...
(by the way, "Update State" is not yet documented, searching Google for the words "budibase" "action" "Update State" and I only found codes that use this action)

The keys appear in your application:
![image](https://user-images.githubusercontent.com/22456825/207993808-f254eaea-47a9-4a86-b43d-d6f3ba770b75.png)

In my app, there are no options for keys:
![image](https://user-images.githubusercontent.com/22456825/207994112-ce329f7d-b9ee-483a-a14a-edc266099375.png)

I created a new application, with only two tables, summarizing your example, and still, it didn't work.
I'm using version 2.2.3, through docker-compose, in Ubuntu.

I tested it in a Windows 10 environment, also using docker-compose, and the result was the same.

Can you help me understand where I'm going wrong?

I am attaching my test application.

[Test-export-1671150458985.tar.gz](https://github.com/Budibase/budibase/files/10241797/Test-export-1671150458985.tar.gz)

**Reply by andz-bb** [reply]
hi @Bitello 

The Update State key dropdown only shows keys that have already been used in the state. If you want to create a new key you just need to type in whatever key name you would like to use to represent the value. Then when you go to work with the state again either in actions or bindings you should see that the key is available. Sorry that you had trouble finding the state docs, the page for App State is here: https://docs.budibase.com/docs/app-state

**Comment by Bitello** [comment]
Fantastic!!!

Santa Claus exists!

These actions are very powerful!

However... I am using a Postgres database and when I try to create a new column of the Formula type I am not presented with the "Formula type" combobox (in BudiDB it is presented)

![image](https://user-images.githubusercontent.com/22456825/208103471-efddeb1a-0ad0-4e81-9df0-47997ab79a70.png)

Also, I can't, in the same way shown in your solution with BudiDB, assign the _id of the row to the formula

![image](https://user-images.githubusercontent.com/22456825/208104347-c0ac0ee5-980d-4e7b-94d0-226f850b068a.png)

Is the given solution restricted to using BudiDB?

**Reply by andz-bb** [reply]
hey @Bitello 🎅

yes the solution was designed with Budibase DB in mind, but using an external SQL table should make this easier. There's no need to create columns that return the foreign key _id as you should already have columns in your tables for that.

The correct solution in your case would be to set the row's id to the state - in your case I think that is `Id` for each table instead of `_id` which is what is used on Budibase DB. Then once that value is set on state you can filter the results on the child table where the Id stored on your state matches the foreign key.

Hope this clears things up but if you would like a specific example for your setup feel free to share your schema and I can demonstrate with that 😊

**Reply by LucianHorrobin** [reply]
Hi andz,
I'm trying to implement this functionality in my app and at first I thought I had it nailed.
I copied your example and then started to play with many to many relationships, to which i discoverd that if a job is assigned to two people, it wont be displayed in the jobs section. Additionally, if the job is removed from the emplyee in my app, the relationship persists as the added column for filtering is static (you cant filter by dynamic formula columns).

So, to overcome these problems, i rebuilt the app using supabase/PosgresSql, though when setting the _id as a state and then filtering the table i get the following error.
![ct](https://github.com/user-attachments/assets/c9863507-38a9-4085-b37a-e29635c0c753)
 
I've tried to figure out how to parse the app state string into an int, and pluck the data, though it is beyond my skills.

Ideally, i'd like to be able to click on the clients table and have the clients assets displayed and inversely click on the assets table and have the clients displayed

There must be a simpler way to achieve filtering by a relationship that i'm missing.

I've attached the app and if you could have a look at it i would be incredibly greatful, or just point me in the right direction.

Thanks

[survey_supabase-export-1720924893193.tar.gz](https://github.com/user-attachments/files/16215433/survey_supabase-export-1720924893193.tar.gz)

**Comment by Bitello** [comment]
Worked perfectly!

Thank you very much!

This solution will be very helpful!
